src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+                                -- block count, but this is fine. See NOTE
+                                -- on reordering corner cases.
+                                over _OldestFirst (Smg.<> pure tip) blockIndices
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices'
+                            , _beaOutValid = BlockApplyFailure
+                            }
+                    AftApplyNonCont -> do
+                        -- Attempt to apply blocks which are not a valid
+                        -- contuniation of our chain.
+                        let
+                            tip  = NE.last (getOldestFirst blockchain)
+                            -- The amount of blocks to skip. One is enough.
+                            skip = 1
+                        blockIndices <- genBlockIndices (tip + 1 + skip)
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices
+                            , _beaOutValid = BlockApplyFailure
+                            }
+            (False, True) -> do
+                -- Success with rollback.
+                len <- getRandomR (1, length blockchain)
+                let
+                    -- 'NE.fromList' is valid here because 'len >= 1'.
+                    select = NE.fromList . NE.take len
+                    blockIndices = toNewestFirst blockchain &
+                        over _NewestFirst select
+                return . BlkEvRollback $ BlockEventRollback
+                    { _berInput    = blockIndices
+                    , _berOutValid = BlockRollbackSuccess
+                    }
+            (False, False) -> do
+                -- Success without rollback (with apply).
+                let tip = NE.last (getOldestFirst blockchain)
+                blockIndices <- genBlockIndices (tip + 1)
+                return . BlkEvApply $ BlockEventApply
+                    { _beaInput = blockIndices
+                    , _beaOutValid = BlockApplySuccess
+                    }
+  where
+    genBlockIndices blockIndexStart = do
+        let blockIndexEndMax = blockIndexStart +
+                               fromIntegral (begp ^. begpBlockCountMax)
+        blockIndexEnd <- getRandomR (blockIndexStart,blockIndexEndMax)
+        return $ OldestFirst . NE.fromList $ [blockIndexStart..blockIndexEnd]
+
+{- NOTE: Reordering corner cases
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When reordering blocks to get an invalid sequence, there's a corner case:
+if there's only one block, it's impossible to get an invalid ordering.
+The reason one might have only one block depends on the means of generation,
+but for 'genBlockIndices' there are two possible causes:
+
+* we rolled 'blockIndexEnd' equal to 'blockIndexStart'
+* 'blockIndexMax' is 1
+
+To avoid dealing with this, we may append one more block. The downside is that
```

Comment:
_we may append one more block_ sounds confusing for me. How should I treat it? It sounds like this is an open question. But apparently that's what actually done.